tests/actions/SessionTest.js

```diff
@@ -0,0 +1,74 @@
+import {fetchAccountDetails, signIn} from '../../src/libs/actions/Session';
+import * as API from '../../src/libs/API';
+import HttpUtils from '../../src/libs/HttpUtils';
+import waitForPromisesToResolve from '../utils/waitForPromisesToResolve';
+
+// Set up manual mocks for methods used in the actions so our test does not fail.
+jest.mock('../../src/libs/Notification/PushNotification', () => ({
+    register: () => {},
+}));
+
+// We are mocking this method so that we can later test to see if it was called and what arguments it was called with.
+// We test HttpUtils.xhr() since this means that our API command turned into a network request and isn't only queued.
+jest.mock('../../src/libs/HttpUtils', () => ({
+    xhr: jest.fn(),
+}));
+
+test('Authenticate is called with saved credentials when a session expires', () => {
+    const TEST_USER_LOGIN = 'test@testguy.com';
+
+    // Set up mock responses for all APIs that will be called. The next time this command is called it will return
+    // jsonCode: 200 and the response here.
+    API.setMockCommand('GetAccountStatus', jest.fn(() => Promise.resolve({
+        jsonCode: 200,
+        accountExists: true,
+        canAccessExpensifyCash: true,
+        requiresTwoFactorAuth: false,
+    })));
+
+    // Simulate user entering their login and populating the credentials.login
+    fetchAccountDetails(TEST_USER_LOGIN);
+
+    // Note: In order for this test to work we must return a promise! It will pass even with
+    // failing assertions if we remove the return keyword.
+    return waitForPromisesToResolve()
+        .then(() => {
+            // Next we will simulate signing in and make sure all API calls in this flow succeed
+            const authenticateMock = jest.fn(() => Promise.resolve({
+                jsonCode: 200,
+                accountID: 1,
+                authToken: '12345',
+                email: TEST_USER_LOGIN,
+            }));
+            API.setMockCommand('Authenticate', authenticateMock);
+            API.setMockCommand('CreateLogin', () => Promise.resolve({
+                jsonCode: 200,
+                accountID: 1,
+                authToken: '12345',
+                email: TEST_USER_LOGIN,
+            }));
+            signIn('Password1');
+            expect(authenticateMock.mock.calls.length).toBe(1);
+            return waitForPromisesToResolve();
+        })
+        .then(() => {
+            // At this point we have an authToken. To simulate it expiring we'll just make another
+            // request and mock the response so it returns 407. Once this happens we should attempt
+            // to Re-Authenticate with the stored credentials. We will also reset all the mocks we've set
+            // so that we can test that Authenticate is getting called via HttpUtils.xhr() and not just
+            // being put into the queue.
+            API.resetMockCommands();
```

Comment:
Since mock commands only live in memory, is there a case where it matters if there is something stale in here? I'm just thinking that this method isn't really necessary.

Reply:
Yeah, this is something I'd like to improve if possible because it's not super obvious. The idea here is that if we don't clear the mock for `Authenticate` then we can't test whether `xhr()` has been called. A mock command would not call `xhr()` but one with no mock would call the command "for real" (but not really for real since `xhr()` itself is mocked). Basically, in this test I am looking to see that `fetch()` will be called without calling `fetch()`.

Reply:
I'll just throw out this alternative as a thinking exercise and see if it might help improve this. Instead of asserting that functions are called (which to me isn't really very valuable information), we make assertions about what is in Onyx. Since Onyx data is the end result that we want, I think it makes more sense to test that rather than how many times functions get called.

I've never writted tests before that test the number of times a method is called though, so I might be mis-understanding some of the value. To me, the output of a function is the valuable piece, not just the knowledge that the function was called. I hope that makes sense.

If we switched to testing the outcome instead of the number of times a method is called, would that clean up these tests and some of the confusion about mocking functions?

Reply:
>If we switched to testing the outcome instead of the number of times a method is called, would that clean up these tests and some of the confusion about mocking functions?

I think it would clear things up. But I think it might just look like us providing mock responses to an `API.SomeCommand()` and then testing that this data made it into `Onyx`. While that definitely protects us from changes to an action e.g. someone goofing up and merging the wrong data. That isn't really the test that I set out to write since the details I want to make more resilient are everything that happens in between calling `API.SomeCommand()` and the `Onyx.merge()`.

Scenarios like the authToken expiring happen in a sort of black box (one that keeps breaking and one we keep tinkering with) so that's why I felt it would be better to only mock the `fetch()` and `Network.post()` response.

That said, if we had an E2E testing set up with a test database we could maybe create these accounts for real. I wasn't too sure whether it would be valuable to get that up and running.

Reply:
WRT to testing the whether a function was called is valuable. I do think that makes a lot of practical sense and agree that in most cases we might not care (especially with Onyx responsible for storing those values and providing a clear way for us to inspect everything).

In this case, I think the only way to prove that `Authenticate` will be called is by mocking the response of `Network.post()` so we can see that `handleExpiredAuthToken()` is called and another request to `Authenticate` was made.

We could maybe also not test if `HttpUtils.xhr()` was called but mock it's response as well with an authToken that we gave it to return and then look for that in `Onyx`. If we did this, I still think it would ultimately be valuable to make sure that the `processNetworkRequest()` flow is getting tested.

Maybe one way to clean this up could be to get rid of `API.setMockResponse()` and instead mock the responses of `Network.post()` and `HttpUtils.xhr()`. But this way made more sense to my brain.

Reply:
Ok I thought about this some more. And I think if we _only_ mock the response of `HttpUtils.xhr()` we can greatly simplify all of this, and still test all the logic I wanted to test.

Reply:
OK yeah, I like this a LOT better and it gets rid of all the `API.setMockResponse()` cruft

https://github.com/Expensify/Expensify.cash/pull/1195/commits/842daab969c5b6ffb834c8983f31b5bad79f0c42

Reply:
Cool, I followed that logic and I can see in those cases why it might be valuable to test that some methods were called. I think there is probably a balance between testing both which methods were called, and also the final outcome. It will certainly protect this flow more going forward, though I think it has the potential to create this problem which I've seen in Chronos tests:

- When a code change causes tests to break, we just update what the tests are doing.

This can lead to someone just changing the tests to match the expected output without really understanding why the change resulted in the tests breaking. However, there are some other tests that I've relied on recently with Expensiworks that were really valuable when I made some changes. I wanted to ensure that my changes could be implemented with no changes to the tests, and that gave me a lot of confidence in the code change itself.

Anyway, I'm not suggesting anything, but wanted to give some of my final thoughts on this thread. 🍿

Reply:
>This can lead to someone just changing the tests to match the expected output without really understanding why the change resulted in the tests breaking

Yeah, those are great insights. I like that tests are check on whether you know what you are doing or not. But fixing a broken test without understanding _why_ it broke undermines that check. With any test there's always a chance someone will modify the test to make it useless. But for certain trouble areas of the codebase I don't mind the extra step to answer - "Does anyone know why this test is breaking? Am I doing something wrong? Does the test need to be updated?".